From ef4a526ec0868c052f151a1a6a4c3031db8c8f74 Mon Sep 17 00:00:00 2001
From: Silvano di Ninno <silvano.dininno@nxp.com>
Date: Tue, 3 Mar 2020 16:36:29 +0100
Subject: [PATCH 56/87] TEE-465: drivers: caam: disable job ring through fdt
 overlay

on i.MX 8 OP-TEE does not have direct access to device tree.
Uses overlay to disable Job Ring node.

Signed-off-by: Silvano di Ninno <silvano.dininno@nxp.com>
---
 .../drivers/crypto/caam/hal/common/hal_cfg_dt.c | 17 +++++++++++++++++
 core/drivers/crypto/caam/include/caam_hal_cfg.h | 10 ++++++++++
 2 files changed, 27 insertions(+)

diff --git a/core/drivers/crypto/caam/hal/common/hal_cfg_dt.c b/core/drivers/crypto/caam/hal/common/hal_cfg_dt.c
index f03a7d54..17a0272e 100644
--- a/core/drivers/crypto/caam/hal/common/hal_cfg_dt.c
+++ b/core/drivers/crypto/caam/hal/common/hal_cfg_dt.c
@@ -8,10 +8,13 @@
 #include <caam_hal_cfg.h>
 #include <caam_hal_jr.h>
 #include <caam_jr.h>
+#include <config.h>
 #include <kernel/dt.h>
+#include <kernel/generic_boot.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
+#include <stdio.h>
 
 static const char *dt_caam_match_table = {
 	"fsl,sec-v4.0",
@@ -133,4 +136,18 @@ void caam_hal_cfg_disable_jobring_dt(void *fdt, struct caam_jrcfg *jrcfg)
 			break;
 		}
 	}
+
+	if (IS_ENABLED(CFG_EXTERNAL_DTB_OVERLAY)) {
+		if (node == -FDT_ERR_NOTFOUND) {
+			char target[64];
+			int ret;
+
+			ret = snprintf(target, sizeof(target),
+				       DTB_JR_PATH "@%lx", jrcfg->offset);
+			if (ret < 0 || (size_t)ret >= sizeof(target))
+				panic();
+			if (dt_overlay_disable_node(target))
+				panic();
+		}
+	}
 }
diff --git a/core/drivers/crypto/caam/include/caam_hal_cfg.h b/core/drivers/crypto/caam/include/caam_hal_cfg.h
index 2d805620..da2cb083 100644
--- a/core/drivers/crypto/caam/include/caam_hal_cfg.h
+++ b/core/drivers/crypto/caam/include/caam_hal_cfg.h
@@ -9,6 +9,16 @@
 
 #include <caam_jr.h>
 
+#if defined(CFG_MX8MM) || defined(CFG_MX8MN) || defined(CFG_MX8MP) ||          \
+	defined(CFG_MX8MQ)
+#define DTB_JR_PATH "/soc@0/bus@30800000/crypto@30900000/jr"
+#elif defined(CFG_MX8QM) || defined(CFG_MX8QXP) || defined(CFG_MX8DXL)
+#define DTB_JR_PATH "/bus@31400000/crypto@31400000/jr"
+#else
+/* Currently not supported on i.MX 6 and 7 */
+#define DTB_JR_PATH ""
+#endif
+
 /*
  * Returns the Job Ring Configuration to be used by the TEE
  *
-- 
2.17.1

